mqtt/init.lua

```diff
@@ -81,6 +81,183 @@ function mqtt.run_sync(cl)
+	end
+	return "^"..t.."$"
+end
+
+--- Parses wildcards in a topic into a table.
+-- Options include:
+--
+-- - `opts.topic`: the wild-carded topic to match against (optional if `opts.pattern` is given)
+--
+-- - `opts.pattern`: the compiled pattern for the wild-carded topic (optional if `opts.topic`
+--   is given). If not given then topic will be compiled and the result will be
+--   stored in this field for future use (cache).
+--
+-- - `opts.keys`: (optional) array of field names. The order must be the same as the
+--   order of the wildcards in `topic`
+--
+-- Returned tables:
+--
+-- - `fields` table: the array part will have the values of the wildcards, in
+--   the order they appeared. The hash part, will have the field names provided
+--   in `opts.keys`, with the values of the corresponding wildcard. If a `#`
+--   wildcard was used, that one will be the last in the table.
+--
+-- - `varargs` table: will only be returned if the wildcard topic contained the
+--   `#` wildcard. The returned table is an array, with all segments that were
+--   matched by the `#` wildcard.
+-- @param topic (string) incoming topic string (required)
+-- @param opts (table) with options (required)
+-- @return fields (table) + varargs (table or nil), or false+err on error.
+-- @usage
+-- local opts = {
+--   topic = "homes/+/+/#",
+--   keys = { "homeid", "roomid", "varargs"},
+-- }
+-- local fields, varargs = topic_match("homes/myhome/living/mainlights/brightness", opts)
+--
+-- print(fields[1], fields.homeid)  -- "myhome myhome"
+-- print(fields[2], fields.roomid)  -- "living living"
+-- print(fields[3], fields.varargs) -- "mainlights/brightness mainlights/brightness"
+--
+-- print(varargs[1]) -- "mainlights"
+-- print(varargs[2]) -- "brightness"
+function mqtt.topic_match(topic, opts)
+	if type(topic) ~= "string" then
+		return false, "expected topic to be a string"
+	end
+	if type(opts) ~= "table" then
+		return false, "expected optionss to be a table"
+	end
+	local pattern = opts.pattern
+	if not pattern then
+		local ptopic = opts.topic
+		if not ptopic then
+			return false, "either 'opts.topic' or 'opts.pattern' must set"
+		end
+		local err
+		pattern, err = mqtt.compile_topic_pattern(ptopic)
+		if not pattern then
+			return false, "failed to compile 'opts.topic' into pattern: "..tostring(err)
+		end
+		-- store/cache compiled pattern for next time
+		opts.pattern = pattern
+	end
+	local values = { topic:match(pattern) }
+	if values[1] == nil then
+		return false, "topic does not match wildcard pattern"
+	end
+	local keys = opts.keys
+	if keys ~= nil then
+		if type(keys) ~= "table" then
+			return false, "expected 'opts.keys' to be a table (array)"
+		end
+		-- we have a table with keys, copy values to fields
+		for i, value in ipairs(values) do
+			local key = keys[i]
+			if key ~= nil then
+				values[key] = value
+			end
+		end
+	end
+	if not pattern:find("%(%.[%-%+]%)%$$") then -- pattern for "#" as last char
```

Comment:
please move that pattern string (and its related strings from compile_topic_pattern function) into the local variables in the init.lua file to look like constants as they are related very implicitly.

Reply:
Done. moved them into narrow scoped module locals.